### PR TITLE
[OCPNODE-553] Add CRD ImageDigestMirrorSet,ImageTagMirrorSet

### PR DESCRIPTION
**Pull request by QiWang19**
**- What I did**
- Add support for ImageDigestMirrorSet,ImageDigestMirrorSet mirror configuration CRDs.
- Reject the update of one of current icsp and new idms.itms resources if both exist to avoid having both icsp and new CRs on the same cluster.

Close card: https://issues.redhat.com/browse/OCPNODE-553
Epic: https://issues.redhat.com/browse/OCPNODE-521 
Dependent PR: https://github.com/openshift/runtime-utils/pull/15

**- How to verify it**

Apply following CR, check the /etc/containers/registries.conf

```
apiVersion: config.openshift.io/v1
kind: ImageTagMirrorSet
metadata:
  name: tag-mirror
spec:
  imageTagMirrors:
  - mirrors:
    - example.io/example/ubi-minimal 
    source: registry.access.redhat.com/ubi8/ubi-minimal
  - mirrors:
    - example.io/example/ubi-minimal 
    source: registry.access.redhat.com/ubi8/ubi-minimal-1
    mirrorSourcePolicy: NeverContactSource
```
/etc/containers/registries.conf
```
unqualified-search-registries = ["registry.access.redhat.com", "docker.io"]
short-name-mode = ""

[[registry]]
  prefix = ""
  location = "registry.access.redhat.com/ubi8/ubi-minimal"

  [[registry.mirror]]
    location = "example.io/example/ubi-minimal"
    pull-from-mirror = "tag-only"

[[registry]]
  prefix = ""
  location = "registry.access.redhat.com/ubi8/ubi-minimal-1"
  blocked = true

  [[registry.mirror]]
    location = "example.io/example/ubi-minimal"
    pull-from-mirror = "tag-only"
```

**- Description for the changelog**

Add support for ImageDigestMirrorSet,ImageDigestMirrorSet CRDs

### CODE REVIEW

File: pkg/controller/container-runtime-config/container_runtime_config_controller_test.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/container_runtime_config_controller_test.go:192]: Shouldn’t there be _some_ tests that the IDMS/ITMS data are correctly propagated? Around `TestRunImageBootstrap`/ some future variant of `TestICSPUpdate`?

File: pkg/controller/container-runtime-config/helpers.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/helpers.go:547]: I don’t understand what this comment is intended to mean.

File: pkg/controller/container-runtime-config/helpers.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/helpers.go:543]: If it is desirable to reject conflicts, shouldn’t it reject also conflicts between IDMS/ITMS?

File: pkg/controller/container-runtime-config/helpers.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/helpers.go:596]: Would it be easy not to duplicate this?

File: pkg/controller/container-runtime-config/helpers.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/helpers.go:604]: See https://github.com/openshift/runtime-utils/pull/15#discussion_r842734297 — I don’t think every single mirror consumer (and there are quite a few) should copy&paste this.

**Review by mtrmac — COMMENTED** [review]
Just an initial pass (I didn’t actually read the added tests).

File: pkg/controller/container-runtime-config/helpers.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/helpers.go]: Yes, I agree... Moved this helper to runtime-utils.

File: pkg/controller/container-runtime-config/helpers.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/helpers.go]: Removed the comment.

I explained the empty string and reject mirror using regex validation are needed for `ImageDigestMirrorSet` that is converted from `ImageContentSourcePolicy`.
normal `ImageDigestMirrorSet` does not need this validation since the API parser does the check. Removed the comment since the ICSP is not processed here.

File: pkg/controller/container-runtime-config/helpers.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/helpers.go]: fixed.

File: pkg/controller/container-runtime-config/helpers.go
Comment (mtrmac) [reply pkg/controller/container-runtime-config/helpers.go]: Ah, that is actually a good point that is worth recording. I’m afraid I didn’t make a connection between the two sentences.

Maybe something vaguely like
“Unlike ImageContentSourcePolicy, we don’t need to reject empty string values here because Image{Digest,Tag}MirrorSet rejects such values in schema validation already”.
?

**Comment by QiWang19** [comment]
@yuqi-zhang  This PR and dependent PR: https://github.com/openshift/runtime-utils/pull/15 will update the registry.conf. And the underlying implementation replaced `sysregistriesv2.Registry.MirrorByDigestOnly` with new per-mirror field `sysregistriesv2.Endpoint.PullFromMirror` in the registry.conf. So I think this PR should also update `isSafeContainerRegistryConfChanges()`.  https://github.com/openshift/machine-config-operator/blob/be77042b147c02ffa0d73d34bec84df49e8eee4b/pkg/daemon/drain.go#L178-L187

Not sure how the safe conditions are defined, but looking at https://bugzilla.redhat.com/show_bug.cgi?id=1943315#c1, we could propose the following as safe?
```go
// Currently, we consider following container registry config changes as safe to skip node drain:
// 1. A new mirror is added to an existing registry that has per-mirror setting `pull-from-mirror=digest-only` or registry level setting `mirror-by-digest-only=true`
// 2. A new registry has been added that has `mirror-by-digest-only=true` or per-mirror setting `pull-from-mirror=digest-only` for all of its mirrors
```

**Review by mtrmac — COMMENTED** [review]
LGTM overall. (FWIW the PR doesn’t currently compile).

**Review by yuqi-zhang — COMMENTED** [review]
> Not sure how the safe conditions are defined, but looking at https://bugzilla.redhat.com/show_bug.cgi?id=1943315#c1, we could propose the following as safe?

Sorry I didn't see this PR before. I am not quite sure about safety. Is there a list of what you can do under the new model and how they are specified in registries.conf?

Also, I think the verify test is failing due to your API bump deprecating some of the older fields.

**Comment by yuqi-zhang** [comment]
There are a few other PRs updating the API though, e.g. https://github.com/openshift/machine-config-operator/pull/3308, so maybe we can leave that part out and rebase

**Comment by mtrmac** [comment]
> @yuqi-zhang @mtrmac This PR and dependent PR: [openshift/runtime-utils#15](https://github.com/openshift/runtime-utils/pull/15) will update the registry.conf. And the runtime-utils PR replaced `sysregistriesv2.Registry.MirrorByDigestOnly` with new per-mirror field `sysregistriesv2.Endpoint.PullFromMirror` to edit the registry.conf. So I think this PR should also update `isSafeContainerRegistryConfChanges()`.
> 
> https://github.com/openshift/machine-config-operator/blob/be77042b147c02ffa0d73d34bec84df49e8eee4b/pkg/daemon/drain.go#L178-L187
> 
> Not sure how the safe conditions are defined

Looking at that bug, intent _of the bug_ seems to have been to reboot on any change that could have changed _which_ image ends up being pulled (notably if it could change how a tag reference is resolved).

> but looking at https://bugzilla.redhat.com/show_bug.cgi?id=1943315#c1, we could propose the following as safe?
> 
> ```go
> // Currently, we consider following container registry config changes as safe to skip node drain:
> // 1. A new mirror is added to an existing registry that has per-mirror setting `pull-from-mirror=digest-only` or registry level setting `mirror-by-digest-only=true`
> // 2. A new registry has been added that has `mirror-by-digest-only=true` or per-mirror setting `pull-from-mirror=digest-only` for all of its mirrors
> ```

That looks fine to me (assuming an update to IDMS/ITMS naming).

(If there are any tag-mirrors added or removed, we would reboot, and thus make sure that the tag mirror change, if it matters for the tag→digest lookup, is applied _immediately_. I’m not sure that this is a promise that OCP, or MCO, can ever make, but it does seems valuable from the simple point of being able to deploy a change and then know that it worked, or broke something (or possibly being able to automatically gradually rollout a change and/or stop such a rollout if it completely breaks nodes), within a _reasonable_ amount of time. Just changing the config without a drain/reboot would make the change effective faster _for images pulled in the future_, but it might cause already-pulled images not to be affected, delaying the true effect of the change and making testing / diagnosis harder.)

**Comment by QiWang19** [comment]
@mtrmac Could you review this? This also vendors the latest commit from https://github.com/openshift/runtime-utils/pull/15

File: install/0000_80_machine-config-operator_03_rbac.yaml
Comment (mtrmac) [inline install/0000_80_machine-config-operator_03_rbac.yaml:33]: Um, how does this relate to the subject of this PR? Isn’t there some lesser privilege that could be used?

`clusterrole.yaml` is already adding access to the ITMS/IDMS data.

File: manifests/machineconfigcontroller/clusterrole.yaml
Comment (mtrmac) [inline manifests/machineconfigcontroller/clusterrole.yaml:23]: Is the write access still necessary? I thought the current plan is that the conversion is going to happen outside of MCO, maybe via https://github.com/openshift/oc/pull/1238 .

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:761]: It’s a bit surprising that MCO is not calling `EditRegistriesConfig` with the raw ICSP data.

It _can_ work that way, of course, but MCO is the most frequent caller of that function, so having it call `EditRegistriesConfig` exactly the same way as the other callers would decrease the risk of MCO drifting away from the other mirror config consumers (like builds).

This code can still do its own conversion or merging for the purposes of `getValidBlockedAndAllowedRegistries` — maybe inside that function. Handling the ICSP/IDMS duality inside that function would also help us avoid needing the three copies of this loop.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go:872]: Non-blocking: This parameter list is getting very impractical.

Maybe there should be a `struct registriesConfig { insecure, registriesBlocked, policyBlocked, allowed, search, IDMS, ITMS, ICSP }` that a lot of the code could pass around as a single unit. Potentially factor out constructing this data out of the main line of `syncImageConfig` into a helper; and maybe make things like `getValidBlockedAndAllowedRegistries` a method (or a constructor?) on this struct.

That might also help respecting, rather than ignoring, the `// nolint: gocyclo` warning.

(Such a refactoring should almost certainly be a separate commit … or, well, maybe a separate PR entirely)

File: pkg/controller/container-runtime-config/helpers.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/helpers.go:542]:
```suggestion
			// If the payload registry doesn't have mirror rules configured for it, then don't add it to the blocked registries list
			// Note that we care only about digest mirrors (and not ImageTagMirrorSet) because the OpenShift release payload only uses digest references.
```

File: pkg/controller/container-runtime-config/helpers_test.go
Comment (mtrmac) [inline pkg/controller/container-runtime-config/helpers_test.go:362]: Remove `icspRules` if they are no longer being tested

File: pkg/daemon/drain.go
Comment (mtrmac) [inline pkg/daemon/drain.go:284]: Why is this correct?

File: pkg/daemon/drain.go
Comment (mtrmac) [inline pkg/daemon/drain.go:288]: AFAICS this could just be `return false`, without needing the variable.

File: pkg/daemon/drain_test.go
Comment (mtrmac) [inline pkg/daemon/drain_test.go:26]: Note to self: I didn’t review this file.

File: pkg/daemon/drain.go
Comment (mtrmac) [inline pkg/daemon/drain.go:243]: These two loops could be integrated … and the condition can probably be simplified. At the very least it would avoid two successive rounds of `searchRegistryMirror` for the same mirrors.

File: test/e2e/ctrcfg_test.go
Comment (mtrmac) [inline test/e2e/ctrcfg_test.go:9]: Note to self: I didn’t review this file.

File: manifests/machineconfigcontroller/clusterrole.yaml
Comment (QiWang19) [reply manifests/machineconfigcontroller/clusterrole.yaml]: oh, no. I forget to cleanup this when dropping the imagedigestmirrorset objects creat.

File: install/0000_80_machine-config-operator_03_rbac.yaml
Comment (QiWang19) [reply install/0000_80_machine-config-operator_03_rbac.yaml]: removed

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go]: I prefer refactor this in a separate pr.

File: pkg/daemon/drain.go
Comment (QiWang19) [reply pkg/daemon/drain.go]: replace this code with checking registry level MirrorByDigestOnly.

File: pkg/controller/container-runtime-config/helpers.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/helpers.go:546]: comment fixed.

**Comment by QiWang19** [comment]
@mtrmac Could you give it another round of review?

**Comment by rphillips** [comment]
/hold

@mtrmac A final review would be great!

**Review by mtrmac — COMMENTED** [review]
/lgtm

(I didn’t really review the test updates.)

**Comment by QiWang19** [comment]
/hold cancel
4.13 branch is open

**Comment by QiWang19** [comment]
@yuqi-zhang could the MCO team review and approve this PR? It is ready to get in.

**Review by yuqi-zhang — COMMENTED** [review]
Did a first pass with some question/comments below

Also, the MCO team is doing pre-merge QE testing for 4.13. Is node team doing pre-merge QE testing? Would someone from node QE be able to verify this new feature and apply QE approval?

File: pkg/daemon/drain.go
Comment (QiWang19) [reply pkg/daemon/drain.go:128]: It is a 1-1 matching, it happens in openshift/runtime-util: https://github.com/openshift/runtime-utils/blob/master/pkg/registries/registries.go#L296. 
Before this patch, set `mirror-by-digest-only=true` for the registry once, it applies to all the mirrors of the original registry. This `pull-from-mirror=digest-only` is configurable at the mirror level. 
So to keep the same behavior, we add `digest-only` to the mirror if the mirror was under the mirror-by-digest-only registry.

File: test/e2e/ctrcfg_test.go
Comment (QiWang19) [reply test/e2e/ctrcfg_test.go:307]: No. No configurations exist for checking this value roll back after the icsp deletion.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (QiWang19) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:758]: Without getting in https://github.com/openshift/kubernetes/pull/1310, the user will not detect the error until they look at the logs.
I think with https://github.com/openshift/kubernetes/pull/1310  stopping updating different crds from the API level, theoretically, it should fail before getting here.
What https://github.com/openshift/kubernetes/pull/1310 did has been discussed https://docs.google.com/document/d/13h6IJn8wlzXdiPMvCWlMEHOXXqEZ9_GYOl02Wldb3z8/edit#heading=h.shj0ebufp0yj, section Questions 1.c. I hope it can pass the ci

File: test/e2e/ctrcfg_test.go
Comment (yuqi-zhang) [reply test/e2e/ctrcfg_test.go:307]: Ah ok, should we remove these commented lines?

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go
Comment (yuqi-zhang) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go:758]: ack, thanks!

File: pkg/daemon/drain.go
Comment (QiWang19) [reply pkg/daemon/drain.go:220]: Add some unit test cases with `mirror-by-digest-only` for the old style. ICSP will no longer generate this old style, but it is still supported when there are hand-crafted registries.conf.

File: test/e2e/ctrcfg_test.go
Comment (QiWang19) [reply test/e2e/ctrcfg_test.go]: ah :sweat_smile: yes, removed.

File: test/e2e/ctrcfg_test.go
Comment (QiWang19) [reply test/e2e/ctrcfg_test.go:50]: This is just to test the ICSP can still work with the addition of new CRDs, before we deprecate the ICSP.

yes, we can add some tests for IDSM/ITMS.

File: pkg/controller/bootstrap/testdata/bootstrap/image-content-source-policy-0.yaml
Comment (yuqi-zhang) [inline pkg/controller/bootstrap/testdata/bootstrap/image-content-source-policy-0.yaml:10]: Hmm, does this actually change anything? Just curious

**Review by yuqi-zhang — APPROVED** [review]
I think this generally makes sense to me, approving, thanks for working on this!

Also going to add a

/hold

for QE pre-merge testing and QE approval.

Doesn't have to be part of this PR, but could you provide some docs on how to use IDMS/ITMS? Or maybe some links to other docs are fine as well, thanks!

File: pkg/controller/bootstrap/testdata/bootstrap/image-content-source-policy-0.yaml
Comment (QiWang19) [reply pkg/controller/bootstrap/testdata/bootstrap/image-content-source-policy-0.yaml:10]: These files are also used in the e2e test. To fix the error I saw in the ci e2e test that is image repository name must be lowercase.

**Comment by rphillips** [comment]
/lgtm 
/hold for qe review

@QiWang19 Please remove the hold when ready.

**Comment by yuqi-zhang** [comment]
gcp-op may require a rebase to master which has some e2e test fixes

**Comment by QiWang19** [comment]
```  
ctrcfg_test.go:283: default configuration value "digest-only" same as values being tested against. Consider updating the test
```
https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/3037/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1615814135679815680/build-log.txt the 
rebased, e2e-gcp-op result shows the test platform has icsp resources. I can keep the icsp e2e test but have to drop idms,items e2e test. Before the icsp CRD gets deprecated, using idms, items on the same cluster where icsp resource exists is prohibited.

**Comment by QiWang19** [comment]
@yuqi-zhang could you retag lgtm?

**Comment by QiWang19** [comment]
@yuqi-zhang can we skip the tests?

**Comment by yuqi-zhang** [comment]
Only `e2e-aws-ovn` is required, is failing, and being retested, so skipping tests won't really help us either way. The other failures don't block the PR

Or did you mean you wanted me to override e2e-aws-ovn?

**Comment by QiWang19** [comment]
Yes,  override e2e-aws-ovn. The e2e-aws-ovn was retested several times and failed. Can we override it?

**Comment by yuqi-zhang** [comment]
The latest commit did have a passing run: https://prow.ci.openshift.org/pr-history/?org=openshift&repo=machine-config-operator&pr=3037

So I think it's most likely ok. Still, I am uncomfortable overriding a required job unless the fix is urgent or directly attributed to it. Let's wait until the end of today, and, if there are no passes, I can look to override

**Pull request merged**
